libCacheSim/include/libCacheSim/cacheObj.h

```diff
@@ -201,97 +199,244 @@ typedef struct cache_obj {
+  return cache_obj;
 }
 
-/** remove the object from the LRU queue (a built-in doubly linked list)
+/** remove the object from the built-in doubly linked list
  *
  * @param head
  * @param tail
  * @param cache_obj
  */
-void remove_obj_from_list(cache_obj_t **head, cache_obj_t **tail,
-                          cache_obj_t *cache_obj);
+static inline void remove_obj_from_list(cache_obj_t **head, cache_obj_t **tail,
+                                        cache_obj_t *cache_obj) {
+  if (head != NULL && cache_obj == *head) {
+    *head = cache_obj->queue.next;
+    if (cache_obj->queue.next != NULL) cache_obj->queue.next->queue.prev = NULL;
+  }
+  if (tail != NULL && cache_obj == *tail) {
+    *tail = cache_obj->queue.prev;
+    if (cache_obj->queue.prev != NULL) cache_obj->queue.prev->queue.next = NULL;
+  }
+
+  if (cache_obj->queue.prev != NULL)
+    cache_obj->queue.prev->queue.next = cache_obj->queue.next;
+
+  if (cache_obj->queue.next != NULL)
+    cache_obj->queue.next->queue.prev = cache_obj->queue.prev;
+
+  cache_obj->queue.prev = NULL;
+  cache_obj->queue.next = NULL;
+}
 
 /**
- * move an object to the tail of the LRU queue (a doubly linked list)
+ * move an object to the tail of the doubly linked list
  * @param head
  * @param tail
  * @param cache_obj
  */
-void move_obj_to_tail(cache_obj_t **head, cache_obj_t **tail,
-                      cache_obj_t *cache_obj);
+static inline void move_obj_to_tail(cache_obj_t **head, cache_obj_t **tail,
+                                    cache_obj_t *cache_obj) {
+  if (*head == *tail) {
+    // the list only has one element
+    DEBUG_ASSERT(cache_obj == *head);
+    DEBUG_ASSERT(cache_obj->queue.next == NULL);
+    DEBUG_ASSERT(cache_obj->queue.prev == NULL);
+    return;
+  }
+  if (cache_obj == *head) {
+    // change head
+    *head = cache_obj->queue.next;
+    cache_obj->queue.next->queue.prev = NULL;
+
+    // move to tail
+    (*tail)->queue.next = cache_obj;
+    cache_obj->queue.next = NULL;
+    cache_obj->queue.prev = *tail;
+    *tail = cache_obj;
+    return;
+  }
+  if (cache_obj == *tail) {
+    return;
+  }
+
+  // bridge list_prev and next
+  cache_obj->queue.prev->queue.next = cache_obj->queue.next;
+  cache_obj->queue.next->queue.prev = cache_obj->queue.prev;
+
+  // handle current tail
+  (*tail)->queue.next = cache_obj;
+
+  // handle this moving object
+  cache_obj->queue.next = NULL;
+  cache_obj->queue.prev = *tail;
+
+  // handle tail
+  *tail = cache_obj;
+}
 
 /**
- * move an object to the head of the LRU queue (a doubly linked list)
+ * move an object to the head of the doubly linked list
  * @param head
  * @param tail
  * @param cache_obj
  */
-void move_obj_to_head(cache_obj_t **head, cache_obj_t **tail,
-                      cache_obj_t *cache_obj);
+static inline void move_obj_to_head(cache_obj_t **head, cache_obj_t **tail,
+                                    cache_obj_t *cache_obj) {
+  DEBUG_ASSERT(head != NULL);
```

Comment:
LGTM! One tiny issue, for crucial check like `head != NULL`, in other places we used `assert` instead of `DEBUG_ASSERT`. Should we unify it?

Reply:
updating